openfisca_core/holders.py

```diff
@@ -112,164 +110,87 @@ def clone(self):
-        dated_holder = self.get_from_cache(period, parameters.get('extra_params'))
-        if dated_holder.array is not None:
-            return dated_holder
+        # Check that the requested period matches period_behavior
+        if not column.period_behavior == PERMANENT:
+            if ((column.period_behavior == MONTH and period.unit != periods.MONTH) or
+                    (column.period_behavior == YEAR and period.unit != periods.YEAR)):
+                raise ValueError('Computation requested with wrong period unit for variable {} ({} instead of {})'.format(
+                    column.name,
+                    period.unit, column.period_behavior))
+        if period.size != 1:
+            raise ValueError('Computation requested for complex period {} for variable {}'.format(
+                period, column.name))
+
+        # First look for a value already cached
+        holder_or_dated_holder = self.get_from_cache(period, parameters.get('extra_params'))
+        if holder_or_dated_holder.array is not None:
+            return holder_or_dated_holder
         assert self._array is None  # self._array should always be None when dated_holder.array is None.
 
+        # Request a computation
         column_start_instant = periods.instant(column.start)
         column_stop_instant = periods.instant(column.end)
         if (column_start_instant is None or column_start_instant <= period.start) \
                 and (column_stop_instant is None or period.start <= column_stop_instant):
             formula_dated_holder = self.formula.compute(period = period, **parameters)
             assert formula_dated_holder is not None
-            if not column.is_permanent:
-                assert accept_other_period or formula_dated_holder.period == period, \
-                    "Requested period {} differs from {} returned by variable {}".format(period,
-                        formula_dated_holder.period, column.name)
             return formula_dated_holder
         array = self.default_array()
         return self.put_in_cache(array, period)
 
     def compute_add(self, period = None, **parameters):
-        dated_holder = self.get_from_cache(period, parameters.get('extra_params'))
-        if dated_holder.array is not None:
-            return dated_holder
+        # Check that the requested period matches period_behavior
+        if self.column.period_behavior == YEAR and period.unit == periods.MONTH:
+            raise ValueError('Computation on period {} impossible on yearly variable {}'.format(
+                period, self.column.name
+                ))
+
+        if self.column.period_behavior == MONTH:
+            variable_period_unit = periods.MONTH
+        elif self.column.period_behavior == YEAR:
+            variable_period_unit = periods.YEAR
+        else:
+            ValueError('compute_add can be used only for yearly or monthly variables.')
```

Comment:
`compute_add` is not a method we want the formula writer to know about.

`Unable to sum date_de_naissance over 2015 : only variables defined monthly or yearly can be summed over time`

?